docs/protocols/xlang_object_graph_spec.md

```diff
@@ -0,0 +1,612 @@
+# Cross language object graph serialization
+
+Fury xlang serialization is an automatic object serialization framework that supports reference and polymorphism.
+Fury will convert an object from/to fury xlang serialization binary format.
+Fury has two core concepts for xlang serialization:
+
+- **Fury xlang binary format**
+- **Framework implemented in different languages to convert object to/from Fury xlang binary format**
+
+The serialization format is a dynamic binary format. The dynamics and reference/polymorphism support make Fury flexible,
+much more easy to use, but
+also introduce more complexities compared to static serialization frameworks. So the format will be more complex.
+
+## Type Systems
+
+### Data Types
+
+- bool: A boolean value (true or false).
+- byte: An 8-bit signed integer.
+- i16: A 16-bit signed integer.
+- i32: A 32-bit signed integer.
+- i64: A 64-bit signed integer.
+- half-float: A 16-bit floating point number.
+- float: A 32-bit floating point number.
+- double: A 64-bit floating point number including NaN and Infinity.
+- string: A text string encoded using Latin1/UTF16/UTF-8 encoding.
+- enum: a data type consisting of a set of named values. Rust enum with non-predefined field values are not supported as
+  an enum
+- list: A sequence of objects.
+- set: An unordered set of unique elements.
+- map: A map of key-value pairs.
+- time types:
+    - Duration: an absolute length of time independent of any calendar/timezone, as a count of seconds and
+      fractions of seconds at nanosecond resolution.
+    - Timestamp: a point in time independent of any calendar/timezone, as a count of seconds and fractions of
+      seconds at nanosecond resolution. The count is relative to an epoch at UTC midnight on January 1, 1970.
+- decimal: exact decimal value represented as an integer value in two's complement.
+- binary: binary data.
+- array type: only allow numeric component. Other arrays will be taken as List. The implementation should support the
+  interoperability between array and list.
+    - array: multiple dimension array which every subarray can have have different size.
+    - int16_array: one dimension int16 array.
+    - int32_array: one dimension int32 array.
+    - int64_array: one dimension int64 array.
+    - half_float_array: one dimension half_float_16 array.
+    - float_array: one dimension float32 array.
+    - double_array: one dimension float64 array.
```

Comment:
How about bool, byte and enum?

Reply:
I missed bool array, will add it later. 
byte array is just binary, seems unnecessary to add it again.

enum array should use list type